src/plugins/intel_gpu/src/plugin/infer_request.cpp

```diff
@@ -912,6 +943,45 @@ void InferRequest::prepare_input(const cldnn::primitive_id& inputName, Blob::Ptr
 
 void InferRequest::prepare_output(const cldnn::primitive_id& outputName, Blob::Ptr& outputBlob) {
     OV_ITT_SCOPED_TASK(itt::domains::intel_gpu_plugin, "InferRequest::prepare_output");
+    const auto node = findOutputByNodeName(outputName);
+    const bool is_static = node && node->get_output_partial_shape(0).is_static();
+    const bool can_use_usm = m_graph->GetEngine()->use_unified_shared_memory();
+
+    if (is_static && can_use_usm) {
+        const auto outputID = m_graph->MapOutputName(outputName);
+        const auto output_layout = m_graph->GetNetwork()->get_node_output_layout(outputID);
+        const auto output_ptr = static_cast<void*>(_outputs[outputName]->buffer());
+        const auto allocation_type = m_graph->GetEngine()->detect_allocation_type(output_ptr);
+        const auto is_usm_host = allocation_type == cldnn::allocation_type::usm_host;
+        const auto has_device_blob = _deviceOutputs.find(outputName) != _deviceOutputs.end();
+        bool can_skip_allocation = false;
+
+        if (has_device_blob) {
+            auto impl = getBlobImpl(_deviceOutputs[outputName]->as<gpu::ClBlob>());
+
+            OPENVINO_ASSERT(impl, str_device_input_unsupported_blob);
+            OPENVINO_ASSERT(impl->is_allocated(), str_output_not_allocated);
+
+            auto impl_mem = impl->getMemory();
+            auto dst_ptr = outputBlob->cbuffer().as<uint8_t*>();
+            // If device output already exists, we can reuse output blob if buffer has usm_host type and points to the same memory,
+            // so we don't need to allocate new memory
+            can_skip_allocation |= same_host_mem(impl_mem, dst_ptr);
+            // Or if blob has any type except usm_host - in that case explicit copy will be performed anyway
+            can_skip_allocation |= impl_mem->get_allocation_type() != cldnn::allocation_type::usm_host;
+        }
+
+        if (!can_skip_allocation) {
+            if (is_usm_host) {
+                _deviceOutputs[outputName] = create_shared_device_blob(outputBlob->getTensorDesc(), output_layout, _outputs[outputName]->buffer().as<void*>());
+            } else {
+                _deviceOutputs[outputName] = create_device_blob(outputBlob->getTensorDesc());
```

Comment:
The code above is almost same for inputs and outputs. Can it be reused?

Reply:
Yes, moved same code to independent function